lib/temporal/worker/reactor.rb

```diff
@@ -0,0 +1,112 @@
+require 'fiber'
+require 'forwardable'
+require 'set'
+require 'singleton'
+require 'temporal/worker/reactor/promise'
+require 'temporal/worker/reactor/command'
+
+module Temporal
+  class Worker
+    class Reactor
+      include Singleton
+      extend SingleForwardable
+
+      def_delegators :instance, :execute, :attach
+
+      def initialize
+        @queue = Queue.new
+        @fibers = Set.new
+        @thread = nil
+        @running = false
+        @mutex = Mutex.new
+      end
+
+      def execute(&block)
+        mutex.synchronize do
+          unless running?
+            @thread = Thread.new { run_loop }
+            @running = true
+          end
+        end
+
+        enqueue { block.call(self) }
+      end
+
+      def async(&block)
+        promise = Promise.new
+
+        enqueue do
+          fiber = Fiber.current
+
+          resolver = lambda do |value = nil|
+            resume(fiber, value)
+          end
+
+          block.call(resolver)
+
+          value = Fiber.yield
+          promise.set(value)
+        end
+
+        promise
+      end
+
+      def await(&block)
+        promise = async(&block)
+        promise.result
+      end
+
+      def attach
+        thread&.join
+      end
+
+      private
+
+      attr_reader :queue, :thread, :fibers, :mutex
+
+      def running?
+        @running
+      end
```

Comment:
In some technical sense this should be behind the lock, but I don't know that it really matters.

Reply:
It's private and the only callsite is already behind a lock. As far as anything inside the Reactor goes — no need for locks since it's concurrent, but still single-threaded